### ISSUE TITLE
Enforce use of variables to define colours

### ISSUE THREAD

**Issue by olliebennett** [causes/scss-lint#337]
The ColorKeyword linter ensures against `color: purple`, preferring `color: #800080`.

However, the [documentation](https://github.com/causes/scss-lint/blob/master/lib/scss_lint/linter/README.md#colorkeyword) for this linter suggests that...

> A better approach is to use global variables like `$color-text-body`.

Can this be enforced using existing linters? If not, could I propose a new linter, `ColorVariable`, requiring variable definitions (`$color-text-body` etc.) in place of colours (eg. `#333`) themselves?

The reasons for preferring variables over direct colour definitions are already outlined in the [ColorKeyword](https://github.com/causes/scss-lint/blob/master/lib/scss_lint/linter/README.md#colorkeyword) linter docs. In addition to the points mentioned, forcing the use of variables can ensure only colours from the project's colour palette are used.

Some additional considerations:
- The new linter could supersede the existing `ColorKeyword` linter.
- The existing `ColorKeyword` linter must avoid reporting variables.
- Any 'config' or 'common' files defining the colour variables should specify `// scss-lint:disable ColorVariable` (this should be documented, as using it would be compulsory to avoid linting the colour definitions themselves).
- A more involved solution could additionally check the definition of the referenced variable within other files (i.e. to ensure the specified variable indeed referenced a valid colour).

**Comment by sds** [comment]
Hey @olliebennett, thanks for suggesting ways we can improve `scss-lint`.

I think there's value in something like what you describe, but done in a different way. 

First of all, the `ColorKeyword` linter still has value for developers who wish to standardize on always using hex codes for colors (the documentation perhaps doesn't capture this reason effectively, but it does exist). This removes the need to address the first two points in your list.

I think what you're actually looking for is the ability to report a lint for any color that is not referenced as part of a variable definition. In this case, I would call it `ColorReference`.

Thus, we would not report lints for references where a color was being stored in a variable:

```scss
$color-text-body: #333;
```

...but we would report lints for references to explicit colors

```scss
p {
  color: #333;
  @include my-color-adjuster(#ccc, 50);
  ... etc ...
}
```

I think this gets to the core of what you are looking for. Let me know if that's not the case.

**Comment by olliebennett** [comment]
Thanks @sds. I agree that `ColorReference` is more succinct, and yes - your example indeed provides what I am looking for.

I can attempt to create a pull request for this if you don't have the bandwidth to tackle it?

**Comment by sds** [comment]
A pull request would be appreciated. Otherwise I'll get to it when I can. 

Thanks!

**Comment by mikeholm** [comment]
My coworker Andy and I just submitted a PR for this called UseVariablesForProperties that lets you specify the props for which you want to use variables. This way, you can enforce variables for 'color', 'font', etc...

**Comment by Anahkiasen** [comment]
I'd love to see that in

**Comment by sds** [comment]
Implemented in #349.

**Comment by olliebennett** [comment]
Great to see work happening on this. While I think the new `VariableForProperties` linter is very useful, it doesn't quite meet the requirements of this issue because one can still use non-standard colours in their code.

```scss
color: #EEE; // now successfully linted
border: 1px solid #EEE; // still not linted
```

Naturally, I don't want to specify a variable for all `border`s (and configuring lint of `border-color` has no effect on the above). Perhaps we still need a color-specific linter to achieve this? Or could/should the new `VariableforProperties` linter be extended to support this?

The same would apply to fonts, where one could require the `font-family` to be a pre-configured, but not worry about font size.

Not sure if the above has been considered? Again, thanks to @sds and @mikeholm for getting it this far!

**Comment by sds** [comment]
Hey @olliebennett,

Thanks for pointing out this discrepancy. You are right that this particular issue as filed has not quite been addressed yet. I was a bit trigger-happy with closing it.

The `VariableForProperty` linter is still valuable on its own, but we still need a linter that enforces using color literals only in variable declarations.

**Comment by sds** [comment]
I've added the `ColorVariable` linter in e72cb23. Thanks for your input, everyone.
